File: src/lib/couchdb/profileService.ts
import { couchdbClient } from './couchdbClient';
import { v4 as uuidv4 } from 'uuid';

export interface Profile {
  _id: string;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  bio: string | null;
  role: 'trainee' | 'trainer' | 'admin';
  updated_at: string;
  created_at: string;
  type: 'profile';
}

export interface ProfileUpdate {
  username?: string;
  full_name?: string;
  avatar_url?: string;
  bio?: string;
  role?: 'trainee' | 'trainer' | 'admin';
}

class ProfileService {
  private static instance: ProfileService;
  private dbName = 'profiles';

  private constructor() {}

  public static getInstance(): ProfileService {
    if (!ProfileService.instance) {
      ProfileService.instance = new ProfileService();
    }
    return ProfileService.instance;
  }

  public async createProfile(
    userId: string,
    username: string,
    fullName?: string,
    avatarUrl?: string,
    bio?: string,
    role: 'trainee' | 'trainer' | 'admin' = 'trainee'
  ): Promise<string> {
    const profile: Profile = {
      _id: userId,
      username,
      full_name: fullName || null,
      avatar_url: avatarUrl || null,
      bio: bio || null,
      role,
      updated_at: new Date().toISOString(),
      created_at: new Date().toISOString(),
      type: 'profile'
    };

    const result = await couchdbClient.createDocument<{ id: string }>(this.dbName, profile);
    return result.id;
  }

  public async getProfile(userId: string): Promise<Profile | null> {
    try {
      return await couchdbClient.getDocument<Profile>(this.dbName, userId);
    } catch (error) {
      console.error('Error fetching profile:', error);
      return null;
    }
  }

  public async updateProfile(userId: string, updates: ProfileUpdate): Promise<boolean> {
    try {
      const profile = await this.getProfile(userId);
      if (!profile) return false;

      const updatedProfile = {
        ...profile,
        ...updates,
        updated_at: new Date().toISOString()
      };

      await couchdbClient.updateDocument<any>(this.dbName, userId, updatedProfile);
      return true;
    } catch (error) {
      console.error('Error updating profile:', error);
      return false;
    }
  }

  public async deleteProfile(userId: string): Promise<boolean> {
    try {
      await couchdbClient.deleteDocument<any>(this.dbName, userId);
      return true;
    } catch (error) {
      console.error('Error deleting profile:', error);
      return false;
    }
  }

  public async getProfilesByRole(role: 'trainee' | 'trainer' | 'admin'): Promise<Profile[]> {
    try {
      const selector = {
        selector: {
          role,
          type: 'profile'
        }
      };
      const result = await couchdbClient.findDocuments<{ docs: Profile[] }>(this.dbName, selector);
      return result.docs;
    } catch (error) {
      console.error('Error fetching profiles by role:', error);
      return [];
    }
  }
}

export const profileService = ProfileService.getInstance();

File: api/database/profile_service.py
from .couchdb_client import couchdb_client
import uuid
from datetime import datetime

class ProfileService:
    def __init__(self):
        self.db_name = 'profiles'
    
    def create_profile(self, user_id, username, full_name=None, avatar_url=None, bio=None, role='trainee'):
        """Create a new user profile"""
        profile = {
            '_id': user_id,
            'username': username,
            'full_name': full_name,
            'avatar_url': avatar_url,
            'bio': bio,
            'role': role,
            'updated_at': datetime.now().isoformat(),
            'created_at': datetime.now().isoformat(),
            'type': 'profile'
        }
        return couchdb_client.create_document(self.db_name, profile)
    
    def get_profile(self, user_id):
        """Get a user profile by ID"""
        return couchdb_client.get_document(self.db_name, user_id)
    
    def update_profile(self, user_id, updates):
        """Update a user profile"""
        return couchdb_client.update_document(self.db_name, user_id, updates)
    
    def delete_profile(self, user_id):
        """Delete a user profile"""
        return couchdb_client.delete_document(self.db_name, user_id)
    
    def get_profiles_by_role(self, role):
        """Get profiles by role"""
        selector = {
            'selector': {
                'role': role,
                'type': 'profile'
            }
        }
        return couchdb_client.find_documents(self.db_name, selector)

profile_service = ProfileService()
